### PR TITLE
fix: type hints within metrics

### PR DESCRIPTION
**Pull request by taharallouche**
Fixed some type hints in the metrics package.

## Tests

- [x] no new tests required
- [ ] new tests added
- [ ] existing tests adjusted

## Documentation

- [x] no documentation changes needed
- [ ] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

File: fairlearn/metrics/_disaggregated_result.py
Comment (romanlutz) [inline fairlearn/metrics/_disaggregated_result.py:186]: Shouldn't this have a default of `None`?
Also, it's missing a docstring section 😮

Comment (taharallouche) [reply fairlearn/metrics/_disaggregated_result.py]: Indeed :eyes: I added both (I actually copied the argument's docstring from `.apply_grouping`, which is the method that uses it inside `.difference` and `.ratio`)

**Review by TamaraAtanasoska — APPROVED** [review]
LGTM, but I will leave it to @romanlutz to check it out again after commenting and merge.

File: fairlearn/metrics/_annotated_metric_function.py
Comment (adrinjalali) [inline fairlearn/metrics/_annotated_metric_function.py:52]: I prefer this:

```suggestion
        positional_argument_names: List[str] | None = None,
        kw_argument_mapping: Dict[str, str] | None = None,
```

you might need a `from __future__ import annotations` at the start of the file

**Review by adrinjalali — COMMENTED** [review]
Same with other `Optional` usages

File: fairlearn/metrics/_annotated_metric_function.py
Comment (taharallouche) [reply fairlearn/metrics/_annotated_metric_function.py:52]: Oh that would be great actually ! :ok_hand: 

I can by the same occasion get rid of the `Dict`, `List` type hints and use plain standard classes `dict`, `list` etc .. (only for the metrics module for now to limit the scope of this PR :sweat_smile: )

Comment (taharallouche) [reply fairlearn/metrics/_annotated_metric_function.py]: @adrinjalali Done :exploding_head:

**Review by adrinjalali — APPROVED** [review]
much nicer!

**Pull request merged**
